M19 Z Resume From Z and Layer Counting

gralco opened the pull request:
- M19 Z and Resume From Z in LCD main menu — resumes a failed print from given Z height (modified [Vince's Solidoodle code](http://www.soliforum.com/topic/5971/firmware-mods-to-rescue-and-resume-failed-prints/) and included an algorithm to compensate for Z-lifts).
- Added layer counting — found using M114 or Menu > Tune > Layer from an LCD

These features must be included together in this PR because merging them would require additional editing.

CptanPanic committed:
Add calculation of difference between auto-correct bed level plane, and measured topo map.

Wackerbarth committed:
Compiler defines USBCON for USB devices

gralco committed:
Initial M19 Z Resume From Z and Layer Counting

gralco committed:
Initial M19 Z Resume From Z and Layer Counting

gralco committed:
Re-enable endstops.

gralco committed:
Re-enable endstops.

thinkyhead commented:
Very nice. You should add a configuration option to turn this on and off, named `RESUME_FEATURE` or something similar, and wrap the new code in `#ifdef RESUME_FEATURE` conditionals. That way Marlin can be built with or without this feature according to user preferences.

gralco committed:
Should be return not break.

gralco committed:
Should be return not break.

tonnico commented:
I like the idea. But messing the plan_set_position is not good.

Could you make a new function for that? we could also use some more functions for it. Maybe one to count the z-layers? This could we use to display the current layer.

Anyway, good work.

gralco commented:
Yes, I'll include both of your ideas, although the Gcode parsing of the resume_feature intersects with some of the same conditions as the layer counting. See diff: https://www.diffchecker.com/zqlknhg5

The function I will create will still need to be called within `plan_buffer_line(...)` though.

gralco committed:
Needed to be inside of plan_set_position(...) not plan_set_position(...).

gralco committed:
Needed to be inside of plan_set_position(...) not plan_set_position(...).

gralco commented:
Actually it should have been inside of `plan_buffer_line(...)` instead of `plan_set_position(...)`, sorry I'm not sure how I messed that one up.

gralco committed:
Created the macros RESUME_FEATURE and TRACK_LAYER and wrapped everything with them.
Created the functions void floor_z(const float &z) and void layer_count(const float &z) which are called inside plan_buffer_line(...).

As per #2179's discussion.

gralco committed:
Created the macros RESUME_FEATURE and TRACK_LAYER and wrapped everything with them.
Created the functions void floor_z(const float &z) and void layer_count(const float &z) which are called inside plan_buffer_line(...).

As per #2179's discussion.

gralco commented:
Check out https://github.com/gralco/Marlin/commit/1f82f32d0df6f66419ea426945d8cef705738eae and tell me what you think.

gralco committed:
Should be int and int3.

gralco committed:
Should be int and int3.

gralco committed:
Include `extern float planner_disabled_below_z;` in Marlin_main.cpp

AnHardt committed:
Z-height fix V2b

here the othe two occourences of `Z_PROBE_OFFSET_FROM_EXTRUDER` are
replaced with `zprobe_zoffset`.

boelle commented:
Please rebase, the code cant be merged as is

thinkyhead committed:
Cleanup of comments & spacing

AnHardt committed:
Add MagoKimbra to the list of contributors

in Development

gralco committed:
Need to include thinkyhead's changes from f18f689c01e822e77aac498f9f7b8ec685538228 in using MSG_N# instead of MSG_F#.

gralco committed:
Need to include thinkyhead's changes from f18f689c01e822e77aac498f9f7b8ec685538228 in using MSG_N# instead of MSG_F#.

gralco commented:
Rebased and resolved conflicts.

Sniffle committed:
z-probe offset fix

verified via 1.0.2 tag that compiles by multiple people,
credit for the fix goes to paulusjacobus for the initial find and
sniffle for the correction so M851 works.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -104,6 +104,7 @@
  * M1   - Same as M0
  * M17  - Enable/Power all stepper motors
  * M18  - Disable all stepper motors; same as M84
+ * M19  - Resume print from current (or given) Z height (disables all movements below the current Z position, a file must be selected to print after executing this M code)
```
I wonder if we can do this by adding a `Z` parameter to `M24` instead, since it is the command for both Print and Resume. But I could see `M19` being useful in cases where the host is printing a file - but some hosts have their own Resume functions, I believe.

gralco replied:
@thinkyhead
Correct, this way (using a separate M19 code) it would cover all cases.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2962,6 +2990,23 @@ inline void gcode_M17() {
+    if(code_seen('Z'))
+    {
+      gcode_get_destination(); // For Z
+      prepare_move();
```
Do you think `st_synchronize()` should be called before sending `M114`?

gralco replied:
@thinkyhead
Yes, there's certainly no problem in having it there, so better be safe than sorry.

thinkyhead replied:
Yes! It seems that without a call to `st_synchronize`, the `M114` command would be processed immediately, before the move just added to the planner even starts.

gralco replied:
@thinkyhead
Ooooh, that makes sense, I think the reason I didn't experience any issues is because when I tested my code I did not have any other movements in queue. Thank you!

thinkyhead replied:
Actually… it occurs to me that Marlin updates the `current_position` to the end of the next planner move just after it sends it to the planner in `prepare_move`. Since `M114` reports that value, it will show the right numbers even without `st_synchronize`.

thinkyhead committed:
Enable LCD_PIN_RESET if it exists

gralco committed:
Add st_synchronize() before the M114 call.

gralco committed:
Add st_synchronize() before the M114 call.

AnHardt committed:
Correct the sign of zprobe_zoffset for G29

Sign flipped with #2167.
See also comments
https://github.com/MarlinFirmware/Marlin/issues/2040#issuecomment-118568355
and following.

thinkyhead committed:
Add extra EOL after M111

thinkyhead committed:
Prevent a hanging SERIAL_ECHO_START in M111

thinkyhead committed:
Call code_seen only once per parameter

thinkyhead committed:
RIGIDBOARD / RIGIDPANEL

thinkyhead committed:
RigidBoard include for pins.h

thinkyhead committed:
RigidBot CONTROLLERFAN_PIN in Configuration_adv.h

thinkyhead committed:
Cleanup of pins_RAMPS_13.h

thinkyhead committed:
Use RIGIDBOT_PANEL for consistency with online docs

thinkyhead committed:
Move note about LiquidTWI2 to Configuration.h

thinkyhead committed:
Use st_get_position_mm where possible

thinkyhead committed:
Remove STEPPER_RESET_FIX option

thinkyhead committed:
Add STEPPER_RESET_PIN support

thinkyhead committed:
Move Y_STOP_PIN for Printrboard with SD

AnHardt committed:
Remove the additional pin variable in Servo.cpp

as sugested by @c-born
inhttps://github.com/MarlinFirmware/Marlin/issues/1885#issuecomment-92618240

AnHardt committed:
Add Servo::move() to servo.cpp

move(pin, angel) - Sequence of attach(pin), write(angel),
                   if DEACTIVATE_SERVOS_AFTER_MOVE is defined waits SERVO_DEACTIVATION_DELAY, than detaches.

As we have jitter on the servos during moves of the steppers, and detaching them improves this behaviour,
the usual sequence to handle a servo movement is:
attach(pin)
write(angel)
delay(until the servo finished the move)
detache()

Here a function to handle the complete sequence.

AnHardt committed:
Introduce DEACTIVATE_SERVOS_AFTER_MOVE in Configuration.h

DEACTIVATE_SERVOS_AFTER_MOVE is ought to replace PROBE_SERVO_DEACTIVATION_DELAY.
But it will work on all servo moves, not only the probe servo moves.

AnHardt committed:
Activate the new servo.move() function

by replacing the sequences (attach, write, delay, detach), and their conditions with the new function in Marlin.main.cpp
and removing the old configuration in the ABL part of Configuration.h

AnHardt committed:
Deactivate DEACTIVATE_SERVOS_AFTER_MOVE by default

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Update language_ru.h

mkile committed:
Update language_ru.h

AnHardt committed:
Corrected Bed Height vs. Bed Topology for verbosity > 3

and cleaned out some tabs.

Everything else is a rebased PR#2017 as proposed by @CptanPanic

gralco committed:
M114 reports the correct position, st_synchronize() not needed here.

gralco committed:
M114 reports the correct position, st_synchronize() not needed here.

thinkyhead committed:
Cleanup and dev notes in Marlin.h

thinkyhead committed:
Require temp sensor for bed thermal protection

thinkyhead committed:
Define extern delta only for delta

AnHardt committed:
Eliminate the further use of the pin-parameter

thinkyhead committed:
Update configs for new servo deactivation

thinkyhead committed:
Update configs formatting

thinkyhead committed:
Remove PID examples from preconfigured configs

thinkyhead committed:
Fix lsDive filename stack overrun

thinkyhead committed:
Move braces in homeaxis to prevent a hanging else

gralco committed:
Resolving more conflicts.

gralco committed:
Resolving more conflicts.

gralco committed:
Fix more conflicts, again

gralco committed:
Bring back "#if ENABLED(SDSUPPORT)"

gralco committed:
Add another "#if ENABLED(SDSUPPORT)", currently resolving issues with "undefined reference to `plan_buffer_line(float const&, float const&, float const&, float const&, float, unsigned char)'"

gralco committed:
Needed to use address-of as the parameter: "&extruder"

gralco committed:
Updated SanityCheck.h

gralco committed:
Missed void Servo::move(int value) definition

gralco committed:
Condition "if (!planner_disabled_below_z || layer_reached)" required for multi-extruder prints

gralco committed:
Prevent accidental resuming from Z=0
